pkg/util/iptree/iptree.go

```diff
@@ -0,0 +1,679 @@
+// Example:
+//
+// 	r := New[int]()
+//
+//	prefixes := []string{
+//		"0.0.0.0/0",
+//		"10.0.0.0/8",
+//		"10.0.0.0/16",
+//		"10.1.0.0/16",
+//		"10.1.1.0/24",
+//		"10.1.244.0/24",
+//		"10.0.0.0/24",
+//		"10.0.0.3/32",
+//		"192.168.0.0/24",
+//		"192.168.0.0/28",
+//		"192.168.129.0/28",
+//	}
+//	for _, k := range prefixes {
+//		r.InsertPrefix(netip.MustParsePrefix(k), 0)
+//	}
+//
+// (*) means the node is not public, is not storing any value
+//
+// 0.0.0.0/0 --- 10.0.0.0/8 --- *10.0.0.0/15 --- 10.0.0.0/16 --- 10.0.0.0/24 --- 10.0.0.3/32
+//  |                                 |
+//  |                                 \ -------- 10.1.0.0/16 --- 10.1.1.0/24
+//  |                                                 |
+//  |	                                               \ ------- 10.1.244.0/24
+//  |
+//  \------ *192.168.0.0/16 --- 192.168.0.0/24 --- 192.168.0.0/28
+//                   |
+//                    \ -------- 192.168.129.0/28
+
+// node is an element of radix tree with a netip.Prefix optimized to store IP prefixes.
+type node[T any] struct {
+	// prefix network CIDR
+	prefix netip.Prefix
+	// public nodes are used to store values
+	public bool
+	val    T
+
+	child [2]*node[T] // binary tree
+}
+
+// mergeChild allow to compress the tree
+// when n has exactly one child and no value
+// p -> n -> b -> c ==> p -> b -> c
+func (n *node[T]) mergeChild() {
+	// public nodes can not be merged
+	if n.public {
+		return
+	}
+	// can not merge if there are two children
+	if n.child[0] != nil &&
+		n.child[1] != nil {
+		return
+	}
+	// can not merge if there are no children
+	if n.child[0] == nil &&
+		n.child[1] == nil {
+		return
+	}
+	// find the child and merge it
+	var child *node[T]
+	if n.child[0] != nil {
+		child = n.child[0]
+	} else if n.child[1] != nil {
+		child = n.child[1]
+	}
+	n.prefix = child.prefix
+	n.public = child.public
+	n.val = child.val
+	n.child = child.child
+	// remove any references from the deleted node
+	// to avoid memory leak
+	child.child[0] = nil
+	child.child[1] = nil
+}
+
+// Tree is a radix tree for IPv4 and IPv6 networks.
+type Tree[T any] struct {
+	rootV4 *node[T]
+	rootV6 *node[T]
+}
+
+// New creates a new Radix Tree for IP addresses.
+func New[T any]() *Tree[T] {
+	return &Tree[T]{
+		rootV4: &node[T]{
+			prefix: netip.PrefixFrom(netip.IPv4Unspecified(), 0),
+		},
+		rootV6: &node[T]{
+			prefix: netip.PrefixFrom(netip.IPv6Unspecified(), 0),
+		},
+	}
+}
+
+// GetPrefix returns the stored value and true if the exact prefix exists in the tree.
+func (t *Tree[T]) GetPrefix(prefix netip.Prefix) (T, bool) {
+	var zeroT T
+
+	n := t.rootV4
+	if prefix.Addr().Is6() {
+		n = t.rootV6
+	}
+	bitPosition := 0
+	// mask the address for sanity
+	address := prefix.Masked().Addr()
+	// we can't check longer than the request mask
+	mask := prefix.Bits()
+	// walk the network bits of the prefix
+	for bitPosition < mask {
```

Comment:
Low Pri: There are so many loops/iterators that looks like this one. Why don't we go with a unified implementation a la visitor with function that loops with a function pointer as an arg that performs what is needed?